File: examples/python/train_tiny.py
from __future__ import annotations

import argparse

import sentinel as S


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("checkpoint", nargs="?", default="tiny_demo.snlm")
    args = parser.parse_args()

    corpus = [
        "the cat sat on the mat",
        "the dog ran in the park",
        "a quick brown fox jumps",
        "sentinel trains causal language models",
        "cuda packs batches for throughput",
        "tiny models learn short patterns",
    ]

    tok = S.BPETokenizer()
    tok.train(corpus, vocab_size=256)

    train = S.LanguageModelDataset.build(corpus, tok, maximum_token_count=48, build_one_hot=False)
    if train.size == 0:
        raise SystemExit("no examples after tokenization")

    model = S.LanguageModel(
        vocabulary_size=tok.vocab_size,
        embedding_dim=64,
        maximum_position_count=64,
        learning_rate=3e-3,
        block_count=2,
        head_count=4,
    )

    if S.cuda_available():
        model.enable_cuda()
        model.set_prefer_flash_attention(True)
        model.enable_cuda_train()
        model.set_activation_checkpoint_mode(S.ActivationCheckpointMode.Off)
        print("train_tiny: CUDA train enabled")
    else:
        print("train_tiny: CUDA unavailable — host OpenMP train")

    model.train(train, epochs=8, batch_size=4, log_every_epochs=2)
    model.save_checkpoint(args.checkpoint, include_optimizer=False)
    safe_path = args.checkpoint + ".safetensors"
    model.save_safetensors(safe_path)

    roundtrip = S.LanguageModel(tok.vocab_size, 64, 64, 3e-3, 2, 4)
    if S.cuda_available():
        roundtrip.enable_cuda()
    roundtrip.load_checkpoint(safe_path)

    print(
        f"train_tiny: examples={train.size} vocab={tok.vocab_size} "
        f"params~{model.parameter_count / 1e6:.4f}M "
        f"avgLoss={model.average_loss(train):.5f} "
        f"safeLoss={roundtrip.average_loss(train):.5f} "
        f"wrote {args.checkpoint} + {safe_path}"
    )


if __name__ == "__main__":
    main()

File: examples/train_tiny.cpp
/// Minimal end-to-end train example: BPE → tiny causal LM → optional CUDA → checkpoint.
///
///   ./sentinel_train_tiny [checkpoint.snlm]
///
/// No external dataset required.

#include "NeuralNet/Cuda/CudaMatmul.hpp"
#include "NeuralNet/Data/LanguageModelDataset.hpp"
#include "NeuralNet/Network/LanguageModel.hpp"
#include "NeuralNet/Optimizers/Adam.hpp"
#include "NeuralNet/Tokenizer/BPETokenizer.hpp"

#include <cstdio>
#include <filesystem>
#include <iostream>
#include <string>
#include <vector>

int main(int argc, char** argv) {
    setvbuf(stdout, nullptr, _IONBF, 0);

    const std::string checkpointPath = argc > 1 ? argv[1] : "tiny_demo.snlm";

    const std::vector<std::string> corpus = {
        "the cat sat on the mat",
        "the dog ran in the park",
        "a quick brown fox jumps",
        "sentinel trains causal language models",
        "cuda packs batches for throughput",
        "tiny models learn short patterns",
    };

    BPETokenizer tokenizer;
    const int vocabSize = 256;
    tokenizer.train(corpus, vocabSize);

    LanguageModelDataset train = LanguageModelDataset::build(corpus, tokenizer, /*maximumTokenCount=*/48, /*buildOneHot=*/false);
    if (train.examples.empty()) {
        std::cerr << "train_tiny: no examples after tokenization\n";
        return 1;
    }

    const int embed = 64;
    const int blocks = 2;
    const int heads = 4;
    const int maxPos = 64;
    LanguageModel model(tokenizer.vocabSize(), embed, maxPos, Adam(3e-3f), blocks, heads);

    if (CudaMatmul::isAvailable()) {
        model.enableCuda();
        model.setCudaPreferFlashAttention(true);
        model.enableCudaTrain();
        model.setActivationCheckpointMode(ActivationCheckpointMode::Off);
        std::cout << "train_tiny: CUDA train enabled\n";
    } else {
        std::cout << "train_tiny: CUDA unavailable — host OpenMP train\n";
    }

    LanguageModelDataset emptyTest;
    model.train(train, emptyTest, /*epochs=*/8, /*logEveryEpochs=*/2, /*batchSize=*/4, /*gradientAccumulationSteps=*/1);
    model.saveCheckpoint(checkpointPath, /*includeOptimizer=*/false);
    const std::string safePath = checkpointPath + ".safetensors";
    model.saveSafeTensors(safePath);
    const std::string tokPath = std::filesystem::path(checkpointPath).replace_extension(".sbpe").string();
    tokenizer.save(tokPath);

    LanguageModel roundtrip(tokenizer.vocabSize(), embed, maxPos, Adam(3e-3f), blocks, heads);
    if (CudaMatmul::isAvailable())
        roundtrip.enableCuda();
    roundtrip.loadCheckpoint(safePath);
    const float safeLoss = roundtrip.averageLoss(train);

    BPETokenizer tokRoundtrip = BPETokenizer::loadFrom(tokPath);
    const std::string probe = "the cat";
    if (tokRoundtrip.encode(probe) != tokenizer.encode(probe)
        || tokRoundtrip.decode(tokenizer.encode(probe)) != tokenizer.decode(tokenizer.encode(probe))) {
        std::cerr << "train_tiny: tokenizer roundtrip mismatch\n";
        return 1;
    }

    const float loss = model.averageLoss(train);
    std::cout << "train_tiny: examples=" << train.size()
              << " vocab=" << tokenizer.vocabSize()
              << " params≈" << (model.parameterElementCount() / 1.0e6) << "M"
              << " avgLoss=" << loss
              << " safeLoss=" << safeLoss
              << " wrote " << checkpointPath << " + " << safePath << " + " << tokPath << "\n";
    return 0;
}
